sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -220,6 +220,19 @@ class SQLQuerySuite extends QueryTest with BeforeAndAfterAll {
       Seq(Row(1,3), Row(2,3), Row(3,3)))
   }
 
+  test("SPARK-6451 SUM agg with code generation") {
+    val originalValue = conf.codegenEnabled
+    setConf(SQLConf.CODEGEN_ENABLED, "true")
+    checkAnswer(
+      sql("SELECT a, SUM(b), SUM(CAST(b as Double)) FROM testData2 GROUP BY a"),
+      Seq(Row(1, 3, 3.0), Row(2, 3, 3.0), Row(3, 3, 3.0)))
+
+    checkAnswer(
+      sql("SELECT  sum('a'), avg('a') FROM testData"),
+      Row(0, null))
+    setConf(SQLConf.CODEGEN_ENABLED, originalValue.toString)      
+  }
+  
```

Comment:
I also have some tests added in https://github.com/yhuai/spark/blob/expressionType/sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala#L107. Can you combine my stuff into your test?  I think we can use the test of "aggregation with codegen" for now. Also, I think it is necessary to check the plan to make sure we indeed have GeneratedAggregate.